llvm/lib/Target/Mips/NanoMipsCompressJumpTables.cpp

```diff
@@ -0,0 +1,163 @@
+//===- NanoMipsCompressJumpTables.cpp - nanoMIPS compress JTs  --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file This file contains a pass that compresses Jump Table entries, whenever
+/// possible.
+//
+//===----------------------------------------------------------------------===//
+
+#include "Mips.h"
+#include "MipsMachineFunction.h"
+#include "MipsSubtarget.h"
+#include "llvm/CodeGen/MachineBasicBlock.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineJumpTableInfo.h"
+
+#include <cmath>
+
+using namespace llvm;
+
+#define NM_COMPRESS_JUMP_TABLES_OPT_NAME                                       \
+  "nanoMIPS compress jump tables optimization pass"
+
+namespace {
+struct NMCompressJumpTables : public MachineFunctionPass {
+  static char ID;
+  const MipsSubtarget *STI;
+  const TargetInstrInfo *TII;
+  MachineFunction *MF;
+  SmallVector<int, 8> BlockInfo;
+  SmallVector<int, 8> BrOffsets;
+
+  int computeBlockSize(MachineBasicBlock &MBB);
+  void scanFunction();
+  bool compressJumpTable(MachineInstr &MI, int Offset);
+
+  NMCompressJumpTables() : MachineFunctionPass(ID) {}
+  StringRef getPassName() const override {
+    return NM_COMPRESS_JUMP_TABLES_OPT_NAME;
+  }
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+};
+} // namespace
+
+char NMCompressJumpTables::ID = 0;
+
+int NMCompressJumpTables::computeBlockSize(MachineBasicBlock &MBB) {
+  // return MBB.size() * 6;
```

Comment:
What about potential expansion from instructions that will have relocations only resolved at link time? eg. call instructions are generally assumed to be 32 bits but if they are out of range the linker will expand them to a load of the address and a JALRC instruction.

Reply:
Also, the `JumpTableDest*` instructions are pseudos at this point and won't be counted.

Reply:
Unfortunately that's not even enough to guarantee it's always correct, as calls that resolve to very far addresses may be split by the linker into an ADDIUPC[48] and a JALRC[16]/JRC[16] so a total of 8 bytes. That would be **very** pessimistic since most instructions on average are more like ~24 bits.

I don't think there's any existing LLVM interface which we can use to tell the maximum potential size of a `MachineInstr`, so we'd have to add such a function  for the maximum expanded size. I guess locally here would probably be okay, but really we should derive `NanoMipsInstrInfo` (which we haven't yet) and implement it there; there are some other instruction query type functions that could be refactored to there.

Reply:
Thank you for the exhaustive explanation! I understand the problem of not having an appropriate interface for the nanomips yet. But since It is a separate issue, do you think it'd be okay to leave here some kind of heuristic, with a TODO comment to enhance it once we have a suitable interface implemented?

Reply:
Yes, fine to implement here.

Reply:
However, after discussions with Dragan, it seems that the linker will be able to handle jump table expansion on its own if the relocations don't fit, so this is safe after all :)